File: python/find_first_and_last_position_of_element_in_sorted_array.py
"""
https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
"""
from typing import List


def searchRange(nums: List[int], target: int) -> List[int]:
    if len(nums) == 0:
        return [-1, -1]
    found = False

    # binary search for the biggest value smaller than the target
    low = 0
    high = len(nums) - 1
    start = -1
    while low <= high:
        mid = (low + high) // 2
        if nums[mid] < target:
            start = mid
            low = mid + 1
        else:
            if nums[mid] == target:
                found = True
            high = mid - 1
    start = max(start + 1, 0)

    # binary search for the smallest value bigger than the target
    low = 0
    high = len(nums) - 1
    end = -1
    while low <= high:
        mid = (low + high) // 2
        if nums[mid] <= target:
            low = mid + 1
        else:
            end = mid
            high = mid - 1
    if end == -1:
        end = len(nums) - 1
    else:
        end -= 1

    if not found:
        return [-1, -1]
    return [start, end]


if __name__ == "__main__":
    assert searchRange([5, 7, 7, 8, 8, 10], 8) == [3, 4]
    assert searchRange([5, 7, 7, 8, 8, 10], 6) == [-1, -1]
    assert searchRange([], 0) == [-1, -1]

File: cpp/find_first_and_last_position_of_element_in_sorted_array.cpp
/**
 * @file find_first_and_last_position_of_element_in_sorted_array.cpp
 * @brief
 * https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 * @date 2025-03-11
 */
#include <algorithm>
#include <cassert>
#include <iterator>
#include <vector>
using namespace std;

vector<int> searchRange(vector<int> const &nums, int target) {
	auto left = lower_bound(nums.cbegin(), nums.cend(), target);
	auto right = upper_bound(nums.cbegin(), nums.cend(), target);
	return left == right
	           ? vector{-1, -1}
	           : vector{
	                 static_cast<int>(distance(nums.cbegin(), left)),
	                 static_cast<int>(distance(nums.cbegin(), --right)),
	             };
}

int main() {
	assert(searchRange({5, 7, 7, 8, 8, 10}, 8) == vector({3, 4}));
	assert(searchRange({5, 7, 7, 8, 8, 10}, 6) == vector({-1, -1}));
	assert(searchRange({1, 1, 1}, 1) == vector({0, 2}));
	assert(searchRange({}, 0) == vector({-1, -1}));
	return 0;
}
